### DISCUSSION TITLE
High CPU usage on rtmp playback

### DISCUSSION THREAD

**Discussion by mariosutalo** [HaishinKit/HaishinKit.swift#1313] [Q&A]
Hi, we are using your ibrary for RTMP playback, we noticed high CPU usage, especially on older devices. For example on Apple TV 4th generation, when bitrate is higher than 4000kbps cpu usage goes to 100%. On newer devices CPU usage is still high, for example on iphone 12 if bitrate is 9000kbps cpu usage goes over 50%. I have used xcode time profile tool to see details of cpu usage, and 6 or more threads are used to handle reading and processing data from rtmp socket and they use most of CPU. I am attaching screenshot from time profiler. I can send you trace file from profiler too if needed. Thank you in advance for your response.
![Screenshot 2023-10-07 at 12 08 31](https://github.com/shogo4405/HaishinKit.swift/assets/11756624/2a3ae775-e680-443c-8498-9a7300786c7c)

**Comment by shogo4405** [comment]
1. Would changing the Optimization Level to be equivalent to RELEASE improve the performance?
1. I don't consider a 50% CPU utilization to be a problem. By the way, I achieved the following results on my end:
    - iPhone 7 Plus + 9000kbps H264/Base + 1920x1080 + PiPView = CPU Usage 25-50%

**Comment by mariosutalo** [comment]
Hi, release optimization makes no significant difference. On older devices like Apple TV 4th generation (from 2015) which still supports latest tvOS version, CPU usage is very high (>90%) when bitrate is over 4000kbps. What I have seen from source code is that you use removeSubrange method from Data class in file https://github.com/shogo4405/HaishinKit.swift/blob/main/Sources/RTMP/RTMPNWSocket.swift in listen method and that removeSubrange tends to be cpu intensive. Maybe that could be the reason for high cpu usage.

**Reply by shogo4405** [reply]
I'm sorry that the proposed solution didn't work. I see, so you're planning to improve it with that approach. I look forward to the pull request (PR).

**Reply by mariosutalo** [reply]
I will try to optimize it, thank you.
